45-topi-antelope/technical_specification.md

```diff
@@ -0,0 +1,83 @@
+# Structured Logging Tools (Topi Antelope)
+**Epic Type:** Implementation Epic
+
+Epic planning and implementation follow the
+[Epic Planning and Marathon SOP](https://docs.ghga-dev.de/main/sops/sop001_epic_planning.html).
+
+**Attention: Please do not put any confidential content here.**
+
+## Scope
+### Outline:
+The aim of this epic is to implement tools that enable consistent structured logging,
+with the result being a set of minimally-invasive, configurable logging tools. The
+logging functionality is not meant to live within providers or protocols within hexkit,
+but rather be provided as a utility similar to the basic correlation ID functionality.
+In addition, `uvicorn` needs to be configured such that its formatting is consistent
+with other logs.
+
+### Included/Required:
+- Logging Tools in Hexkit
+- Consolidating Uvicorn Logging
+- Pilot Repository
+
+### Not Included:
+- Updating services to use new logging tools.
+
+
+## Additional Implementation Details:
+
+### Logging Tools in `Hexkit`
+There needs to be a way to easily invoke a logger object with minimal boilerplate code.
+There should also be a dedicated logging configuration class that subclasses `BaseSettings`.
+
+The configuration should at least include the log level for the service.
+When employing loggers in a service, it should be as easy as calling a function with a logger
+name as an argument and then using the object to log as needed.
+The logs should be emitted in a JSON string format (one line). An example log message is as
+follows (formatted for ease of reading):
+
+```json
+{
+	"timestamp": "2023-12-04T15:30:00Z",
+	"service": "ucs",
+	"instance": "001",
+	"level": "ERROR",
+	"correlation_id": "d826361b-3734-4590-b3e8-3cbed68b9236",
+	"message": "The file with ID 123xyz is already in the inbox.",
+    "details": {
+		"file_id": "123xyz"
+	}
+}
+```
+Additional configurable properties could include a way to disable uvicorn integration
+(e.g. for easier reading during development), or a format string that can be used to override
+the JSON output format.
```

Comment:
I like that. If it is simple to implement, we could also have a json_log_format boolean parameter in the config to switch between JSON and non-JSON (more readable) logging.

Reply:
That sounds reasonable to me